sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -582,34 +598,59 @@ public void setup() throws Exception {
 
       @StartBundle
       public void startBundle() {
-        batchCount = 0;
+        // nothing to do
       }
 
       @ProcessElement
       public void processElement(ProcessContext context) throws Exception {
         T record = context.element();
 
-        preparedStatement.clearParameters();
-        spec.getPreparedStatementSetter().setParameters(record, preparedStatement);
-        preparedStatement.addBatch();
-
-        batchCount++;
+        records.add(record);
 
-        if (batchCount >= spec.getBatchSize()) {
+        if (records.size() >= spec.getBatchSize()) {
           executeBatch();
         }
       }
 
+      public void processRecord(T record) {
+        try {
+          preparedStatement.clearParameters();
+          spec.getPreparedStatementSetter().setParameters(record, preparedStatement);
+          preparedStatement.addBatch();
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      }
+
       @FinishBundle
       public void finishBundle() throws Exception {
         executeBatch();
       }
 
-      private void executeBatch() throws SQLException {
-        if (batchCount > 0) {
-          preparedStatement.executeBatch();
-          connection.commit();
-          batchCount = 0;
+      private void executeBatch() throws SQLException, IOException, InterruptedException {
+        if (records.size() > 0) {
+          Sleeper sleeper = Sleeper.DEFAULT;
+          BackOff backoff = BUNDLE_WRITE_BACKOFF.backoff();
+          while (true) {
+            try {
+              // add each record in the statement batch
+              records.stream().forEach(this::processRecord);
+              // execute the batch
+              preparedStatement.executeBatch();
+              // commit the changes
+              connection.commit();
+              break;
+            } catch (SQLException exception) {
```

Comment:
Is it possible to detect whether this is a deadlock? Does `SQLException` contain such information in a portable way?

Reply:
Unfortunately, it would depend to vendor error. There's no generic way to detect the `SQLException` is related to a deadlock. We would need to add parsing the `SQLException` depending of the vendor (for instance, detect a `ORA-00060` for Oracle, a `1205` for SQLServer).